File: internal/contract/contract.go
package contract

import (
	"encoding/json"
	"time"
)

const (
	EventMessageSent       = "aether/message.sent"
	EventScheduleFired     = "aether/schedule.fired"
	EventApprovalAnswered  = "aether/approval.answered"
	EventMessageReplied    = "daemon/message.replied"
	EventApprovalRequested = "daemon/approval.requested"
	EventConjureRequested  = "daemon/conjure.requested"
	EventDeployRequested   = "daemon/deploy.requested"
	EventDismissRequested  = "daemon/dismiss.requested"
	EventChangeRequested   = "daemon/change.requested"
)

const (
	TurnConcurrencyKey   = "event.data.thread"
	TurnConcurrencyLimit = 1
)

type Call struct {
	ID      string          `json:"id"`
	Tool    string          `json:"tool"`
	Args    json.RawMessage `json:"args"`
	Context json.RawMessage `json:"context"`
}

type MessageSentPayload struct {
	Daemon  string `json:"daemon"`
	Thread  string `json:"thread"`
	Message string `json:"message"`
	Text    string `json:"text"`
}

type ScheduleFiredPayload struct {
	Daemon     string    `json:"daemon"`
	Thread     string    `json:"thread"`
	Schedule   string    `json:"schedule"`
	DueAt      time.Time `json:"due_at"`
	DeadlineAt time.Time `json:"deadline_at"`
}

type ApprovalAnsweredPayload struct {
	Daemon   string `json:"daemon"`
	Thread   string `json:"thread"`
	Call     string `json:"call"`
	Approval string `json:"approval"`
	Decision string `json:"decision"`
}

type MessageRepliedPayload struct {
	Daemon string `json:"daemon"`
	Thread string `json:"thread"`
	Reply  string `json:"reply"`
	Text   string `json:"text"`
}

type ApprovalRequestedPayload struct {
	Daemon string `json:"daemon"`
	Thread string `json:"thread"`
	Calls  []Call `json:"calls"`
}

type ConjureRequestedPayload struct {
	Daemon   string `json:"daemon"`
	Host     string `json:"host"`
	Language string `json:"language"`
}

type DeployRequestedPayload struct {
	Daemon string `json:"daemon"`
	Host   string `json:"host"`
}

type DismissRequestedPayload struct {
	Daemon string `json:"daemon"`
	Host   string `json:"host"`
}

type ChangeRequestedPayload struct {
	Daemon string   `json:"daemon"`
	Patch  string   `json:"patch"`
	Tests  []string `json:"tests"`
}

var Payloads = map[string]any{
	EventMessageSent:       MessageSentPayload{},
	EventScheduleFired:     ScheduleFiredPayload{},
	EventApprovalAnswered:  ApprovalAnsweredPayload{},
	EventMessageReplied:    MessageRepliedPayload{},
	EventApprovalRequested: ApprovalRequestedPayload{},
	EventConjureRequested:  ConjureRequestedPayload{},
	EventDeployRequested:   DeployRequestedPayload{},
	EventDismissRequested:  DismissRequestedPayload{},
	EventChangeRequested:   ChangeRequestedPayload{},
}

File: packages/daemon/src/contract.ts

export const Events = {
  MessageSent: "aether/message.sent",
  ScheduleFired: "aether/schedule.fired",
  ApprovalAnswered: "aether/approval.answered",
  MessageReplied: "daemon/message.replied",
  ApprovalRequested: "daemon/approval.requested",
  ConjureRequested: "daemon/conjure.requested",
  DeployRequested: "daemon/deploy.requested",
  DismissRequested: "daemon/dismiss.requested",
  ChangeRequested: "daemon/change.requested",
} as const;

export type EventName = (typeof Events)[keyof typeof Events];

export interface ToolCall {
  id: string;
  tool: string;
  args: Record<string, unknown>;
  context: {
    cwd: string;
    host: string;
  };
}

export interface MessageSentPayload {
  daemon: string;
  thread: string;
  message: string;
  text: string;
}

export interface ScheduleFiredPayload {
  daemon: string;
  thread: string;
  schedule: string;
  due_at: string;
  deadline_at: string;
  prompt: string;
}

export interface Decision {
  call: string;
  approval: string;
  decision: "approved" | "denied";
}

export interface ApprovalAnsweredPayload {
  daemon: string;
  thread: string;
  group: string;
  decisions: Decision[];
}

export interface MessageRepliedPayload {
  daemon: string;
  thread: string;
  reply: string;
  text: string;
}

export interface ApprovalRequestedPayload {
  daemon: string;
  thread: string;
  calls: ToolCall[];
}

export interface ConjureRequestedPayload {
  daemon: string;
  host: string;
  language: string;
}

export interface DeployRequestedPayload {
  daemon: string;
  host: string;
}

export interface DismissRequestedPayload {
  daemon: string;
  host: string;
}

export interface ChangeRequestedPayload {
  daemon: string;
  patch: string;
  tests: string[];
}

export type Payloads = {
  [Events.MessageSent]: MessageSentPayload;
  [Events.ScheduleFired]: ScheduleFiredPayload;
  [Events.ApprovalAnswered]: ApprovalAnsweredPayload;
  [Events.MessageReplied]: MessageRepliedPayload;
  [Events.ApprovalRequested]: ApprovalRequestedPayload;
  [Events.ConjureRequested]: ConjureRequestedPayload;
  [Events.DeployRequested]: DeployRequestedPayload;
  [Events.DismissRequested]: DismissRequestedPayload;
  [Events.ChangeRequested]: ChangeRequestedPayload;
};
export const TurnConcurrency = {
  key: "event.data.thread",
  limit: 1,
} as const;
